src/runtime/hexagon_remote/qurt_thread_pool.cpp

```diff
@@ -0,0 +1,271 @@
+
+extern "C" {
+
+#include "bin/src/halide_hexagon_remote.h"
+#include <memory.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <qurt.h>
+
+}
+
+#include "../HalideRuntime.h"
+
+#define MAX_WORKER_THREADS 4
+#define NUM_WORKER_THREADS_TO_CREATE (MAX_WORKER_THREADS - 1)
+#define STACK_SIZE 256 * 1024
+
+static char stack [MAX_WORKER_THREADS][STACK_SIZE] __attribute__ ((__aligned__(128)));
+qurt_sem_t wait_for_work;
+
+//qurt_cond_t work_in_queue;
+struct work {
+    work *next_job;
+    int (*f)(void *, int, uint8_t *);
+    void *user_context;
+    int next, end;
+    uint8_t *closure;
+    int active_workers;
+    int exit_status;
+    qurt_cond_t wakeup_owner;
+    qurt_hvx_mode_t curr_hvx_mode;
+    // A job can be in the following states.
+    // claimed - When the thread_pool has started work on the entire job, but not necessarily
+    //           completed it. Condition: next >= end.
+    // running - When there are active workers on the job. Condition: active_workers > 0;
+    // done - When the job is completely done and there are no active workers on the job.
+    bool claimed() { return next >= end; }
+    bool running() { return active_workers > 0; }
+    bool done() { return claimed() && !running(); }
+};
+
+bool thread_pool_initialized = false;
+struct work_queue_t {
+    // all fields are protected by this mutex.
+    qurt_mutex_t work_mutex;
+
+    // Jobs that the thread pool needs to work on.
+    work *jobs;
+
+    // Global flag indicating if the thread pool has been shutdown.
+    bool shutdown;
+
+    bool running() {
+        return !shutdown;
+    }
+    void lock() {
+        qurt_mutex_lock(&work_mutex);
+    }
+    void unlock() {
+        qurt_mutex_unlock(&work_mutex);
+    }
+    void init() {
+        qurt_mutex_init(&work_mutex);
+        lock();
+        shutdown = false;
+        jobs = NULL;
+        unlock();
+    }
+};
+work_queue_t work_queue;
+qurt_thread_t threads[MAX_WORKER_THREADS];
+
+// This function does the real work of the thread pool.
+// owned_job is used to tell the differece between the master thread
+// and the worker threads. If owned_job is non-null then it means that
+// we are in the master thread, i.e. the thread that put owned_job in
+// the work queue. This thread, like other worker threads in this function,
+// tries to acquire a lock on the work queue. Each thread then looks for
+// a job to do. In the case of the master thread, it is "owned_job" while
+// a worker thread has to iterate backwards over the work queue and find
+// a job to work on. Once a job has been found, the thread, after some
+// book-keeping, releases the lock and call halide_do_task to do the job.
+// After the job is done, the lock on the queue is re-acquired to update
+// the status and the thread loops again to look for new work. If a worker
+// thread doesn't find any new work, it goes to sleep until awoken by
+// the master thread.
+void worker_thread(work *owned_job) {
```

Comment:
This routine would likely be better if it were split into an inner routine that does the common work for owned and non-owned cases and two separate routines for owned control and non-owned.